File: sections/42_advanced_data_structures_mixed/problems/d_nested_ranges_count/solution.py
import bisect
import sys


class Fenwick:
    def __init__(self, size: int) -> None:
        self.bit = [0] * (size + 1)

    def add(self, index: int, delta: int) -> None:
        while index < len(self.bit):
            self.bit[index] += delta
            index += index & -index

    def prefix_sum(self, index: int) -> int:
        total = 0
        while index:
            total += self.bit[index]
            index -= index & -index
        return total


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    ranges = [(data[2 * i + 1], data[2 * i + 2], i) for i in range(n)]
    rights = sorted({right for _, right, _ in ranges})
    ranges.sort(key=lambda interval: (interval[0], -interval[1]))

    def rank_of(right: int) -> int:
        return bisect.bisect_left(rights, right) + 1

    contains = [0] * n
    contained_by = [0] * n

    bit = Fenwick(len(rights))
    processed = 0
    start = 0
    while start < n:
        end = start + 1
        while end < n and ranges[end][:2] == ranges[start][:2]:
            end += 1
        group_size = end - start
        rank = rank_of(ranges[start][1])
        base = processed - bit.prefix_sum(rank - 1)
        for _, _, index in ranges[start:end]:
            contained_by[index] = base + group_size - 1
        bit.add(rank, group_size)
        processed += group_size
        start = end

    bit = Fenwick(len(rights))
    end = n
    while end:
        start = end - 1
        while start > 0 and ranges[start - 1][:2] == ranges[end - 1][:2]:
            start -= 1
        group_size = end - start
        rank = rank_of(ranges[start][1])
        base = bit.prefix_sum(rank)
        for _, _, index in ranges[start:end]:
            contains[index] = base + group_size - 1
        bit.add(rank, group_size)
        end = start

    print(" ".join(map(str, contains)))
    print(" ".join(map(str, contained_by)))


if __name__ == "__main__":
    main()

File: sections/42_advanced_data_structures_mixed/problems/d_nested_ranges_count/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    vector<int> bit;
    explicit Fenwick(int n) : bit(n + 1, 0) {}

    void add(int index, int delta) {
        for (int n = (int)bit.size(); index < n; index += index & -index) {
            bit[index] += delta;
        }
    }

    int sum(int index) const {
        int result = 0;
        for (; index > 0; index -= index & -index) result += bit[index];
        return result;
    }
};

struct Interval {
    long long left, right;
    int index;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<Interval> ranges(n);
    vector<long long> rights;
    for (int i = 0; i < n; i++) {
        cin >> ranges[i].left >> ranges[i].right;
        ranges[i].index = i;
        rights.push_back(ranges[i].right);
    }
    sort(rights.begin(), rights.end());
    rights.erase(unique(rights.begin(), rights.end()), rights.end());
    sort(ranges.begin(), ranges.end(), [](const Interval& a, const Interval& b) {
        if (a.left != b.left) return a.left < b.left;
        return a.right > b.right;
    });
    auto rank_of = [&](long long right) {
        return int(lower_bound(rights.begin(), rights.end(), right) - rights.begin()) + 1;
    };

    vector<int> contains(n), contained_by(n);
    Fenwick bit((int)rights.size());
    int processed = 0;
    for (int start = 0; start < n;) {
        int end = start + 1;
        while (end < n && ranges[end].left == ranges[start].left &&
               ranges[end].right == ranges[start].right) {
            end++;
        }
        int size = end - start;
        int rank = rank_of(ranges[start].right);
        int base = processed - bit.sum(rank - 1);
        for (int i = start; i < end; i++) contained_by[ranges[i].index] = base + size - 1;
        bit.add(rank, size);
        processed += size;
        start = end;
    }

    bit = Fenwick((int)rights.size());
    for (int end = n; end > 0;) {
        int start = end - 1;
        while (start > 0 && ranges[start - 1].left == ranges[end - 1].left &&
               ranges[start - 1].right == ranges[end - 1].right) {
            start--;
        }
        int size = end - start;
        int rank = rank_of(ranges[start].right);
        int base = bit.sum(rank);
        for (int i = start; i < end; i++) contains[ranges[i].index] = base + size - 1;
        bit.add(rank, size);
        end = start;
    }

    for (int i = 0; i < n; i++) cout << contains[i] << (i + 1 == n ? '\n' : ' ');
    for (int i = 0; i < n; i++) cout << contained_by[i] << (i + 1 == n ? '\n' : ' ');
    return 0;
}
